File: 04-hash-tables/02-collision-resolution/c/open_addressing.c
/**
 * 開放定址法雜湊表（Open Addressing Hash Table）- C 語言實作
 * 支援線性探測、平方探測和雙重雜湊
 *
 * Open Addressing Hash Table with linear probing, quadratic probing, and double hashing
 */

#include "open_addressing.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

// ========== 私有函式 Private Functions ==========

/**
 * 主雜湊函式（使用 djb2 演算法）
 * Primary hash function using djb2 algorithm
 *
 * @param key 要雜湊的字串
 * @param capacity 容量
 * @return 雜湊值
 */
static size_t hash1(const char* key, size_t capacity) {
    unsigned long hash = 5381;
    int c;

    while ((c = *key++)) {
        // hash * 33 + c
        hash = ((hash << 5) + hash) + c;
    }

    return hash % capacity;
}

/**
 * 次雜湊函式（用於雙重雜湊）
 * Secondary hash function for double hashing
 *
 * @param key 要雜湊的字串
 * @param capacity 容量
 * @return 雜湊值（必須非零）
 */
static size_t hash2(const char* key, size_t capacity) {
    unsigned long hash = 0;
    int c;

    while ((c = *key++)) {
        hash = hash * 31 + c;
    }

    // 確保結果非零且為奇數（避免與偶數容量產生問題）
    // Ensure result is non-zero and odd (avoids issues with even capacity)
    size_t result = (hash % (capacity - 1)) + 1;
    return result | 1;  // 確保為奇數 - Ensure odd
}

/**
 * 計算探測位置
 * Calculate probe position
 *
 * @param ht 雜湊表指標
 * @param key 鍵
 * @param i 探測次數（從 0 開始）
 * @return 探測位置
 */
static size_t probe(OpenAddressingHashTable* ht, const char* key, size_t i) {
    size_t h1 = hash1(key, ht->capacity);

    switch (ht->probe_method) {
        case PROBE_LINEAR:
            // 線性探測: h(k, i) = (h1(k) + i) mod m
            // Linear probing: h(k, i) = (h1(k) + i) mod m
            return (h1 + i) % ht->capacity;

        case PROBE_QUADRATIC:
            // 平方探測: h(k, i) = (h1(k) + c1*i + c2*i²) mod m
            // Quadratic probing: h(k, i) = (h1(k) + c1*i + c2*i²) mod m
            // 使用 c1=1, c2=1
            return (h1 + i + i * i) % ht->capacity;

        case PROBE_DOUBLE_HASH:
            // 雙重雜湊: h(k, i) = (h1(k) + i*h2(k)) mod m
            // Double hashing: h(k, i) = (h1(k) + i*h2(k)) mod m
            {
                size_t h2 = hash2(key, ht->capacity);
                return (h1 + i * h2) % ht->capacity;
            }

        default:
            return h1;
    }
}

/**
 * 檢查條目是否為空（未使用）
 * Check if entry is empty (unused)
 */
static bool is_empty(OpenAddressingEntry* entry) {
    return entry->key == NULL && entry->value == NULL;
}

/**
 * 檢查條目是否為墓碑
 * Check if entry is a tombstone
 */
static bool is_tombstone(OpenAddressingEntry* entry) {
    return entry->key == NULL && entry->value == OAHT_TOMBSTONE;
}

/**
 * 檢查條目是否為有效元素
 * Check if entry is a valid element
 */
static bool is_valid(OpenAddressingEntry* entry) {
    return entry->key != NULL;
}

/**
 * 擴容並重新雜湊
 * Resize and rehash
 */
static bool rehash(OpenAddressingHashTable* ht) {
    size_t old_capacity = ht->capacity;
    OpenAddressingEntry* old_entries = ht->entries;

    // 容量加倍 - Double capacity
    size_t new_capacity = old_capacity * 2;
    OpenAddressingEntry* new_entries = calloc(new_capacity, sizeof(OpenAddressingEntry));
    if (!new_entries) return false;

    // 更新雜湊表 - Update hash table
    ht->entries = new_entries;
    ht->capacity = new_capacity;
    size_t old_size = ht->size;
    ht->size = 0;
    ht->tombstones = 0;

    // 重新插入所有有效元素 - Reinsert all valid elements
    for (size_t i = 0; i < old_capacity; i++) {
        OpenAddressingEntry* entry = &old_entries[i];
        if (is_valid(entry)) {
            oaht_insert(ht, entry->key, entry->value);
            free(entry->key);
        }
    }

    free(old_entries);
    return true;
}

// ========== 公開函式 Public Functions ==========

OpenAddressingHashTable* oaht_create(size_t capacity, ProbeMethod method) {
    if (capacity == 0) {
        capacity = OAHT_DEFAULT_CAPACITY;
    }

    OpenAddressingHashTable* ht = malloc(sizeof(OpenAddressingHashTable));
    if (!ht) return NULL;

    ht->entries = calloc(capacity, sizeof(OpenAddressingEntry));
    if (!ht->entries) {
        free(ht);
        return NULL;
    }

    ht->capacity = capacity;
    ht->size = 0;
    ht->tombstones = 0;
    ht->probe_method = method;
    return ht;
}

void oaht_free(OpenAddressingHashTable* ht) {
    if (!ht) return;

    // 釋放所有 key 字串 - Free all key strings
    for (size_t i = 0; i < ht->capacity; i++) {
        if (is_valid(&ht->entries[i])) {
            free(ht->entries[i].key);
        }
    }

    free(ht->entries);
    free(ht);
}

bool oaht_insert(OpenAddressingHashTable* ht, const char* key, void* value) {
    if (!ht || !key || !value || value == OAHT_TOMBSTONE) return false;

    // 檢查是否需要擴容 - Check if rehashing needed
    double effective_load = (double)(ht->size + ht->tombstones) / ht->capacity;
    if (effective_load > OAHT_MAX_LOAD_FACTOR) {
        rehash(ht);
    }

    size_t first_tombstone = SIZE_MAX;

    // 探測找到插入位置 - Probe to find insertion position
    for (size_t i = 0; i < ht->capacity; i++) {
        size_t index = probe(ht, key, i);
        OpenAddressingEntry* entry = &ht->entries[index];

        if (is_empty(entry)) {
            // 找到空位 - Found empty slot
            if (first_tombstone != SIZE_MAX) {
                // 優先使用之前遇到的墓碑位置
                // Use previously encountered tombstone position
                index = first_tombstone;
                entry = &ht->entries[index];
                ht->tombstones--;
            }

            entry->key = strdup(key);
            if (!entry->key) return false;
            entry->value = value;
            ht->size++;
            return true;
        } else if (is_tombstone(entry)) {
            // 記錄第一個墓碑位置 - Record first tombstone position
            if (first_tombstone == SIZE_MAX) {
                first_tombstone = index;
            }
        } else if (strcmp(entry->key, key) == 0) {
            // 鍵已存在，更新值 - Key exists, update value
            entry->value = value;
            return true;
        }
    }

    // 表已滿（理論上不會發生，因為有擴容機制）
    // Table is full (should not happen due to rehashing)
    return false;
}

void* oaht_search(OpenAddressingHashTable* ht, const char* key) {
    if (!ht || !key) return NULL;

    // 探測搜尋 - Probe to search
    for (size_t i = 0; i < ht->capacity; i++) {
        size_t index = probe(ht, key, i);
        OpenAddressingEntry* entry = &ht->entries[index];

        if (is_empty(entry)) {
            // 遇到空位，搜尋失敗 - Found empty slot, search failed
            return NULL;
        } else if (is_valid(entry) && strcmp(entry->key, key) == 0) {
            // 找到 - Found
            return entry->value;
        }
        // 遇到墓碑或其他鍵，繼續探測 - Found tombstone or other key, continue probing
    }

    return NULL;
}

bool oaht_delete(OpenAddressingHashTable* ht, const char* key) {
    if (!ht || !key) return false;

    // 探測搜尋 - Probe to search
    for (size_t i = 0; i < ht->capacity; i++) {
        size_t index = probe(ht, key, i);
        OpenAddressingEntry* entry = &ht->entries[index];

        if (is_empty(entry)) {
            // 遇到空位，刪除失敗 - Found empty slot, deletion failed
            return false;
        } else if (is_valid(entry) && strcmp(entry->key, key) == 0) {
            // 找到，標記為墓碑 - Found, mark as tombstone
            free(entry->key);
            entry->key = NULL;
            entry->value = OAHT_TOMBSTONE;
            ht->size--;
            ht->tombstones++;
            return true;
        }
        // 遇到墓碑或其他鍵，繼續探測 - Found tombstone or other key, continue probing
    }

    return false;
}

bool oaht_contains(OpenAddressingHashTable* ht, const char* key) {
    return oaht_search(ht, key) != NULL;
}

size_t oaht_size(OpenAddressingHashTable* ht) {
    return ht ? ht->size : 0;
}

size_t oaht_capacity(OpenAddressingHashTable* ht) {
    return ht ? ht->capacity : 0;
}

size_t oaht_tombstone_count(OpenAddressingHashTable* ht) {
    return ht ? ht->tombstones : 0;
}

bool oaht_is_empty(OpenAddressingHashTable* ht) {
    return ht ? ht->size == 0 : true;
}

double oaht_load_factor(OpenAddressingHashTable* ht) {
    if (!ht || ht->capacity == 0) return 0.0;
    return (double)(ht->size + ht->tombstones) / ht->capacity;
}

void oaht_clear(OpenAddressingHashTable* ht) {
    if (!ht) return;

    for (size_t i = 0; i < ht->capacity; i++) {
        OpenAddressingEntry* entry = &ht->entries[i];
        if (is_valid(entry)) {
            free(entry->key);
        }
        entry->key = NULL;
        entry->value = NULL;
    }
    ht->size = 0;
    ht->tombstones = 0;
}

const char* oaht_probe_method_name(ProbeMethod method) {
    switch (method) {
        case PROBE_LINEAR:
            return "Linear Probing 線性探測";
        case PROBE_QUADRATIC:
            return "Quadratic Probing 平方探測";
        case PROBE_DOUBLE_HASH:
            return "Double Hashing 雙重雜湊";
        default:
            return "Unknown 未知";
    }
}

void oaht_print(OpenAddressingHashTable* ht) {
    if (!ht) {
        printf("OpenAddressingHashTable: NULL\n");
        return;
    }

    printf("OpenAddressingHashTable (size=%zu, capacity=%zu, load=%.3f, method=%s):\n",
           ht->size, ht->capacity, oaht_load_factor(ht),
           oaht_probe_method_name(ht->probe_method));

    for (size_t i = 0; i < ht->capacity; i++) {
        OpenAddressingEntry* entry = &ht->entries[i];
        if (is_valid(entry)) {
            printf("  [%zu]: (%s, %p)\n", i, entry->key, entry->value);
        } else if (is_tombstone(entry)) {
            printf("  [%zu]: TOMBSTONE\n", i);
        }
    }
}

void oaht_print_stats(OpenAddressingHashTable* ht) {
    if (!ht) {
        printf("OpenAddressingHashTable: NULL\n");
        return;
    }

    size_t empty_slots = 0;
    for (size_t i = 0; i < ht->capacity; i++) {
        if (is_empty(&ht->entries[i])) {
            empty_slots++;
        }
    }

    printf("開放定址法雜湊表統計 Open Addressing Hash Table Statistics:\n");
    printf("  探測方法 Probe Method: %s\n", oaht_probe_method_name(ht->probe_method));
    printf("  元素數量 Size: %zu\n", ht->size);
    printf("  容量 Capacity: %zu\n", ht->capacity);
    printf("  墓碑數量 Tombstones: %zu\n", ht->tombstones);
    printf("  空槽數量 Empty Slots: %zu\n", empty_slots);
    printf("  負載因子 Load Factor: %.3f\n", oaht_load_factor(ht));
    printf("  實際負載 Actual Load: %.3f (不含墓碑 excluding tombstones)\n",
           ht->capacity > 0 ? (double)ht->size / ht->capacity : 0.0);
}

File: 04-hash-tables/02-collision-resolution/python/open_addressing.py
"""
碰撞處理：開放定址法（Open Addressing）- Python 實作
Collision Resolution: Open Addressing implementation

實作三種探測方法：線性探測、二次探測、雙重雜湊
Implements three probing methods: Linear, Quadratic, Double Hashing
"""

from typing import Any, Optional, Tuple
from enum import Enum


class ProbeMethod(Enum):
    """探測方法列舉 - Probing method enum"""
    LINEAR = "linear"           # 線性探測
    QUADRATIC = "quadratic"     # 二次探測
    DOUBLE_HASH = "double_hash" # 雙重雜湊


class OpenAddressingHashTable:
    """
    使用開放定址法的雜湊表
    Hash table using open addressing for collision resolution
    """

    # 特殊標記 - Special markers
    EMPTY = object()      # 空槽位
    DELETED = object()    # 墓碑標記（已刪除）

    DEFAULT_CAPACITY = 16
    MAX_LOAD_FACTOR = 0.7  # 開放定址法需要較低的負載因子

    def __init__(self, capacity: int = DEFAULT_CAPACITY,
                 probe_method: ProbeMethod = ProbeMethod.LINEAR) -> None:
        """
        初始化雜湊表
        Initialize hash table

        Args:
            capacity: 容量（建議使用質數）
            probe_method: 探測方法
        """
        if capacity <= 0:
            raise ValueError("容量必須為正整數 / Capacity must be positive")

        self._capacity = capacity
        self._size = 0
        self._probe_method = probe_method
        # 使用兩個陣列分別存 key 和 value
        self._keys = [self.EMPTY] * capacity
        self._values = [None] * capacity
        # 統計資料 - Statistics
        self._total_probes = 0
        self._operations = 0

    def _hash1(self, key: Any) -> int:
        """主雜湊函數 - Primary hash function"""
        return hash(key) % self._capacity

    def _hash2(self, key: Any) -> int:
        """
        次雜湊函數（用於雙重雜湊）
        Secondary hash function for double hashing

        確保回傳值與 capacity 互質
        Ensures return value is coprime with capacity
        """
        # 使用 capacity - 1 確保結果 >= 1
        return 1 + (hash(key) % (self._capacity - 1))

    def _probe(self, key: Any, i: int) -> int:
        """
        計算第 i 次探測的索引
        Calculate index for i-th probe

        Args:
            key: 鍵
            i: 探測次數（從 0 開始）

        Returns:
            探測索引
        """
        h1 = self._hash1(key)

        if self._probe_method == ProbeMethod.LINEAR:
            # 線性探測：h(k) + i
            return (h1 + i) % self._capacity

        elif self._probe_method == ProbeMethod.QUADRATIC:
            # 二次探測：h(k) + i²
            return (h1 + i * i) % self._capacity

        elif self._probe_method == ProbeMethod.DOUBLE_HASH:
            # 雙重雜湊：h₁(k) + i * h₂(k)
            h2 = self._hash2(key)
            return (h1 + i * h2) % self._capacity

        return h1

    def insert(self, key: Any, value: Any) -> Tuple[bool, int]:
        """
        插入鍵值對
        Insert key-value pair

        Returns:
            (success, probes) - 是否成功與探測次數
        """
        if self.load_factor >= self.MAX_LOAD_FACTOR:
            self._rehash()

        first_deleted = -1  # 記錄第一個墓碑位置

        for i in range(self._capacity):
            index = self._probe(key, i)

            if self._keys[index] is self.EMPTY:
                # 找到空位 - Found empty slot
                if first_deleted != -1:
                    index = first_deleted  # 使用之前的墓碑位置
                self._keys[index] = key
                self._values[index] = value
                self._size += 1
                self._total_probes += i + 1
                self._operations += 1
                return True, i + 1

            elif self._keys[index] is self.DELETED:
                # 記錄第一個墓碑 - Record first tombstone
                if first_deleted == -1:
                    first_deleted = index

            elif self._keys[index] == key:
                # 鍵已存在，更新值 - Key exists, update value
                self._values[index] = value
                self._total_probes += i + 1
                self._operations += 1
                return True, i + 1

        # 表滿（不應發生，因為有負載因子限制）
        self._operations += 1
        return False, self._capacity

    def search(self, key: Any) -> Tuple[Optional[Any], int]:
        """
        搜尋鍵對應的值
        Search for value by key

        Returns:
            (value, probes) - 值與探測次數，未找到回傳 (None, probes)
        """
        for i in range(self._capacity):
            index = self._probe(key, i)

            if self._keys[index] is self.EMPTY:
                # 遇到空位，鍵不存在 - Empty slot, key doesn't exist
                self._total_probes += i + 1
                self._operations += 1
                return None, i + 1

            elif self._keys[index] is self.DELETED:
                # 跳過墓碑繼續探測 - Skip tombstone, continue probing
                continue

            elif self._keys[index] == key:
                # 找到鍵 - Found key
                self._total_probes += i + 1
                self._operations += 1
                return self._values[index], i + 1

        self._total_probes += self._capacity
        self._operations += 1
        return None, self._capacity

    def delete(self, key: Any) -> Tuple[bool, int]:
        """
        刪除鍵值對（使用墓碑標記）
        Delete key-value pair (using tombstone marker)

        Returns:
            (success, probes) - 是否成功與探測次數
        """
        for i in range(self._capacity):
            index = self._probe(key, i)

            if self._keys[index] is self.EMPTY:
                # 遇到空位，鍵不存在 - Empty slot, key doesn't exist
                self._total_probes += i + 1
                self._operations += 1
                return False, i + 1

            elif self._keys[index] is self.DELETED:
                # 跳過墓碑 - Skip tombstone
                continue

            elif self._keys[index] == key:
                # 找到鍵，標記為已刪除 - Found key, mark as deleted
                self._keys[index] = self.DELETED
                self._values[index] = None
                self._size -= 1
                self._total_probes += i + 1
                self._operations += 1
                return True, i + 1

        self._total_probes += self._capacity
        self._operations += 1
        return False, self._capacity

    def _rehash(self) -> None:
        """擴容並重新雜湊 - Resize and rehash"""
        old_keys = self._keys
        old_values = self._values
        old_capacity = self._capacity

        # 容量加倍 - Double capacity
        self._capacity *= 2
        self._keys = [self.EMPTY] * self._capacity
        self._values = [None] * self._capacity
        self._size = 0

        # 重新插入（跳過 EMPTY 和 DELETED）
        for i in range(old_capacity):
            if old_keys[i] is not self.EMPTY and old_keys[i] is not self.DELETED:
                self.insert(old_keys[i], old_values[i])

    @property
    def load_factor(self) -> float:
        """負載因子 - Load factor"""
        return self._size / self._capacity

    @property
    def size(self) -> int:
        return self._size

    @property
    def capacity(self) -> int:
        return self._capacity

    @property
    def probe_method(self) -> ProbeMethod:
        return self._probe_method

    def average_probes(self) -> float:
        """平均探測次數 - Average number of probes"""
        if self._operations == 0:
            return 0.0
        return self._total_probes / self._operations

    def __len__(self) -> int:
        return self._size

    def __contains__(self, key: Any) -> bool:
        value, _ = self.search(key)
        return value is not None

    def visualize(self) -> str:
        """視覺化雜湊表狀態 - Visualize hash table state"""
        result = []
        for i in range(self._capacity):
            if self._keys[i] is self.EMPTY:
                result.append(f"[{i}]: EMPTY")
            elif self._keys[i] is self.DELETED:
                result.append(f"[{i}]: DELETED")
            else:
                result.append(f"[{i}]: {self._keys[i]!r} -> {self._values[i]!r}")
        return "\n".join(result)


# ========== 比較不同探測方法 Compare Probing Methods ==========

def compare_probing_methods():
    """比較三種探測方法的效能 - Compare performance of probing methods"""
    import random

    print("=== 探測方法比較 Probing Methods Comparison ===\n")

    # 產生測試資料 - Generate test data
    random.seed(42)
    test_keys = [f"key_{i}" for i in range(50)]
    random.shuffle(test_keys)

    for method in ProbeMethod:
        print(f"\n--- {method.value.upper()} ---")
        ht = OpenAddressingHashTable(capacity=67, probe_method=method)

        # 插入 - Insert
        for i, key in enumerate(test_keys):
            ht.insert(key, i)

        print(f"元素數量: {len(ht)}")
        print(f"容量: {ht.capacity}")
        print(f"負載因子: {ht.load_factor:.3f}")
        print(f"平均探測次數: {ht.average_probes():.3f}")

        # 搜尋測試 - Search test
        ht._total_probes = 0
        ht._operations = 0
        for key in test_keys[:20]:
            ht.search(key)
        print(f"搜尋平均探測: {ht.average_probes():.3f}")


if __name__ == "__main__":
    print("=== 開放定址法雜湊表示範 Open Addressing Hash Table Demo ===\n")

    # 使用線性探測 - Using linear probing
    ht = OpenAddressingHashTable(capacity=11, probe_method=ProbeMethod.LINEAR)

    # 插入資料 - Insert data
    keys = ["A", "B", "C", "D", "E", "F"]
    print("插入資料 Inserting data:")
    for i, key in enumerate(keys):
        success, probes = ht.insert(key, i * 10)
        print(f"  insert('{key}') - 成功: {success}, 探測: {probes}")

    print(f"\n雜湊表狀態 Hash table state:")
    print(ht.visualize())

    print(f"\n統計 Statistics:")
    print(f"  負載因子: {ht.load_factor:.3f}")
    print(f"  平均探測: {ht.average_probes():.3f}")

    # 刪除測試 - Delete test
    print("\n刪除 'C' Delete 'C':")
    success, probes = ht.delete("C")
    print(f"  成功: {success}, 探測: {probes}")
    print(f"\n刪除後狀態 After deletion:")
    print(ht.visualize())

    # 搜尋測試 - Search test
    print("\n搜尋測試 Search test:")
    for key in ["A", "C", "F", "Z"]:
        value, probes = ht.search(key)
        print(f"  search('{key}') = {value}, 探測: {probes}")

    # 比較探測方法 - Compare probing methods
    print("\n" + "="*50)
    compare_probing_methods()
